src/sensors.cpp

```diff
@@ -1,261 +1,201 @@
-    t_output = esp_timer_get_time();
-    if (t_output - output_intr_times[1] > 10) {
-        portENTER_CRITICAL_ISR(&output_mux);
-        output_intr_times[0] = output_intr_times[1];
-        output_intr_times[1] = t_output;
-        portEXIT_CRITICAL_ISR(&output_mux);
+static esp_err_t IRAM_ATTR read_and_reset_pcnt(pcnt_unit_handle_t unit, int* dest) {
+    esp_err_t res = pcnt_unit_get_count(unit, dest);
+    if (ESP_OK == res) {
+        pcnt_unit_clear_count(unit);
     }
+    return res;
 }
 
-static intr_handle_t rpm_timer_handle;
-static void IRAM_ATTR on_rpm_timer(void* args) {
-    // Clear timer interrupt
-    TIMERG0.int_clr_timers.t0 = 1;
-    TIMERG0.hw_timer[0].config.alarm_en = 1;
-    uint64_t time_delta_n2 = 0;
-    uint64_t time_delta_n3 = 0;
-    uint64_t time_delta_output = 0;
-    portENTER_CRITICAL_ISR(&n2_mux);
-    time_delta_n2 = n2_intr_times[1] - n2_intr_times[0];
-    portEXIT_CRITICAL_ISR(&n2_mux);
-    portENTER_CRITICAL_ISR(&n3_mux);
-    time_delta_n3 = n3_intr_times[1] - n3_intr_times[0];
-    portEXIT_CRITICAL_ISR(&n3_mux);
-    portENTER_CRITICAL_ISR(&output_mux);
-    time_delta_output = output_intr_times[1] - output_intr_times[0];
-    portEXIT_CRITICAL_ISR(&output_mux);
+static bool IRAM_ATTR on_rpm_timer(gptimer_handle_t timer, const gptimer_alarm_event_data_t *edata, void *user_data) {
+    int pulses = 0;
+    // N2 Sensor
+    if (ESP_OK == read_and_reset_pcnt(PCNT_HANDLE_N2, &pulses)) {
+        n2_total = n2_total - n2_avgs[avg_n2_idx];
+        n2_avgs[avg_n2_idx] = pulses;
+        n2_total = n2_total + pulses;
+        avg_n2_idx = (avg_n2_idx+1)%RPM_SAMPLES_DEBOUNCE;
+    } else {
 
-    if (time_delta_n3 != 0) {
-        n3_total -= n3_avgs[avg_n3_idx];
-        n3_avgs[avg_n3_idx] = time_delta_n3;
-        n3_total += time_delta_n3;
-        avg_n3_idx = (avg_n3_idx+1)%RPM_SAMPLES_DEBOUNCE;
     }
+    // N3 Sensor
+    if (ESP_OK == read_and_reset_pcnt(PCNT_HANDLE_N3, &pulses)) {
+        n3_total = n3_total - n3_avgs[avg_n3_idx];
+        n3_avgs[avg_n3_idx] = pulses;
+        n3_total = n3_total + pulses;
+        avg_n3_idx = (avg_n3_idx+1)%RPM_SAMPLES_DEBOUNCE;
+    } else {
 
-    if (time_delta_n2 != 0) {
-        n2_total -= n2_avgs[avg_n2_idx];
-        n2_avgs[avg_n2_idx] = time_delta_n2;
-        n2_total += time_delta_n2;
-        avg_n2_idx = (avg_n2_idx+1)%RPM_SAMPLES_DEBOUNCE;
     }
-
-    if (time_delta_output != 0) {
-        output_total -= output_avgs[avg_out_idx];
-        output_avgs[avg_out_idx] = time_delta_output;
-        output_total += time_delta_output;
+    // Output Sensor (If present)
+    if (output_rpm_ok && ESP_OK == read_and_reset_pcnt(PCNT_HANDLE_OUTPUT, &pulses)) {
+        output_total = output_total - output_avgs[avg_out_idx];
+        output_avgs[avg_out_idx] = pulses;
+        output_total = output_total + pulses;
         avg_out_idx = (avg_out_idx+1)%RPM_SAMPLES_DEBOUNCE;
+    } else {
+
     }
+    return true;
```

Comment:
Since the function always returns true, it could also be of type void.

Reply:
See the docs [here](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/gptimer.html), the return true is needed in this ISR context.